MaxText/kernels/ragged_attention.py

```diff
@@ -0,0 +1,302 @@
+  """Ragged multi query attention.
+
+  Args:
+    q: A [batch_size, 1, head_dim] jax.Array.
+    k: A [batch_size, seq_len, head_dim] jax.Array.
+    v: A [batch_size, seq_len, head_dim] jax.Array.
+    lengths: A i32[batch_size] jax.Array.
+    mask_value: The value used for padding in attention. By default it is a very
+      negative floating point number.
+    cost_estimate: A Pallas TPU cost estimate based on a reference implementation
+
+  Returns:
+    The output of attention([batch_size, num_heads, head_dim]), along with the
+    max logit ([batch_size, num_heads, 1]) and softmax denominator ([batch_size,
+    num_heads, 1]).
+  """
+  batch_size, num_heads, head_dim = q.shape 
+  assert lengths.shape == (batch_size,)
+  assert lengths.dtype == jnp.int32
+  seq_len = k.shape[1]  
+
+  def compute_ragged_block_indices(b, i, lengths_ref):
+    length = lengths_ref[b]
+    not_done = i * block_size < length
+    am_last_batch = b == batch_size - 1
+    last_good_block = lax.div(length, block_size) - 1
+    b_next = jnp.where(not_done, b, jnp.where(am_last_batch, b, b + 1))
+    i_next = jnp.where(not_done, i, jnp.where(am_last_batch, last_good_block, 0))
+    return b_next, i_next, 0
+
+  out, m, l = pl.pallas_call(
+      functools.partial(
+          ragged_flash_attention_kernel,
+          block_size=block_size,
+          mask_value=mask_value,
+      ),
+      grid_spec=pltpu.PrefetchScalarGridSpec(
+          num_scalar_prefetch=1,
+          in_specs=[
+              pl.BlockSpec(
+                  (None, num_heads, head_dim),
+                  lambda b, i, _: (b, 0, 0)),
+              pl.BlockSpec(
+                  (None, block_size, head_dim),       
+                  compute_ragged_block_indices),
+              pl.BlockSpec(
+                  (None, block_size, head_dim),       
+                  compute_ragged_block_indices),
+          ],
+          out_specs=[
+              pl.BlockSpec(
+                  (None, num_heads, head_dim),
+                  lambda b, i, _: (b, 0, 0)),
+              pl.BlockSpec(
+                  (None, num_heads, head_dim),
+                  lambda b, i, _: (b, 0, 0)),
+              pl.BlockSpec(
+                  (None, num_heads, head_dim),
+                  lambda b, i, _: (b, 0, 0)),
+          ],
+          grid=(batch_size, seq_len // block_size),
+      ),
+      compiler_params=dict(
+        mosaic=dict(
+          dimension_semantics=("parallel", "arbitrary"),
+          cost_estimate=cost_estimate,
+        )
+      ),
+      out_shape=[
+          jax.ShapeDtypeStruct((batch_size, num_heads, head_dim), jnp.float32),
+          jax.ShapeDtypeStruct((batch_size, num_heads, head_dim), jnp.float32),
+          jax.ShapeDtypeStruct((batch_size, num_heads, head_dim), jnp.float32),
+      ],
+  )(lengths, q, k, v)
+  return out, m[..., 0], l[..., 0]
+
+
+@functools.partial(
+  jax.jit,
+  static_argnames=[
+    "block_size",
+    "mask_value",
+  ],
+)
+def ragged_mha(
```

Comment:
I only see mqa and mha, is ragged gqa also supported?

Reply:
GQA was not in the original PR but is now included.